cdc/log.cc

```diff
@@ -416,6 +416,194 @@ utils::UUID generate_timeuuid(api::timestamp_type t) {
+    clustering_key key;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+};
+
+struct clustered_row_update {
+    gc_clock::rep ttl;
+    clustering_key key;
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+struct clustered_row_deletion {
+    clustering_key key;
+};
+
+struct clustered_range_deletion {
+    clustering_key_prefix start;
+    bound_kind start_kind;
+    clustering_key_prefix end;
+    bound_kind end_kind;
+};
+
+struct batch {
+    std::vector<static_row_update> static_updates;
+    std::vector<clustered_row_insert> clustered_inserts;
+    std::vector<clustered_row_update> clustered_updates;
+    std::vector<clustered_row_deletion> clustered_row_deletions;
+    std::vector<clustered_range_deletion> clustered_range_deletions;
+    bool has_partition_deletion = false;
+};
+
+struct row_update {
+    std::vector<atomic_column_update> atomic_entries;
+    std::vector<nonatomic_column_deletion> nonatomic_deletions;
+    std::vector<nonatomic_column_update> nonatomic_updates;
+};
+
+using set_of_changes = std::map<api::timestamp_type, batch>;
+
+set_of_changes extract_changes(const mutation& base_mutation, const schema& base_schema) {
+    auto extract_row_updates = [&base_schema] (const row& r, column_kind ckind) {
+        std::map<std::pair<api::timestamp_type, gc_clock::rep>, row_update> ups;
+        r.for_each_cell([&] (column_id id, const atomic_cell_or_collection& cell) {
+            auto& cdef = base_schema.column_at(ckind, id);
+            if (cdef.is_atomic()) {
+                auto view = cell.as_atomic_cell(cdef);
+                auto timestamp_and_ttl = std::make_pair(
+                        view.timestamp(),
+                        view.is_live_and_has_ttl() ? view.ttl().count() : 0
+                    );
+                ups[timestamp_and_ttl].atomic_entries.push_back({id, atomic_cell(*cdef.type, view)});
+                return;
+            }
+
+            cell.as_collection_mutation().with_deserialized(*cdef.type, [&] (collection_mutation_view_description mview) {
+                auto desc = mview.materialize(*cdef.type);
+                for (auto& [k, v]: desc.cells) {
+                    auto timestamp_and_ttl = std::make_pair(
+                            v.timestamp(),
+                            v.is_live_and_has_ttl() ? v.ttl().count() : 0
+                        );
+                    auto& updates = ups[timestamp_and_ttl].nonatomic_updates;
+                    if (updates.empty() || updates.back().id != id) {
+                        updates.push_back({id, {}});
+                    }
+                    updates.back().cells.push_back({std::move(k), std::move(v)});
+                }
+
+                auto timestamp_and_ttl = std::make_pair(desc.tomb.timestamp, 0);
```

Comment:
No need for `make_pair`, we have class template deduction now, so `std::pair()` works too.

Reply:
The tomb might be missing, in which case you don't need to store it.

Reply:
true